src/crafting.cpp

```diff
@@ -1914,51 +1968,85 @@ void player::consume_tools(const std::vector<tool_comp> &tools, int batch, const
             if (map_inv.has_charges(type, count)) {
                 map_has.push_back(*it);
             }
-        } else if (has_amount(type, 1) || map_inv.has_tools(type, 1)) {
+        }
+        else if (has_amount(type, 1) || map_inv.has_tools(type, 1)) {
             found_nocharge = true;
         }
     }
     if (found_nocharge) {
-        return;    // Default to using a tool that doesn't require charges
+        selected.use_from = usage::use_from_none;
+        return selected;    // Default to using a tool that doesn't require charges
     }
 
     if (player_has.size() + map_has.size() == 1) {
-        if(map_has.empty()) {
-            use_charges(player_has[0].type, player_has[0].count * batch);
+        if (map_has.empty()) {
+            selected.use_from = usage::use_from_player;
+            selected.comp = player_has[0];
         } else {
-            long quantity = map_has[0].count * batch;
-            g->m.use_charges(pos3(), PICKUP_RANGE, map_has[0].type, quantity);
+            selected.use_from = usage::use_from_map;
+            selected.comp = map_has[0];
         }
     } else { // Variety of options, list them and pick one
-        // Populate the list
+           // Populate the list
         std::vector<std::string> options;
-        for( auto &map_ha : map_has ) {
-            std::string tmpStr = item::nname( map_ha.type ) + _( " (nearby)" );
+        for (auto &map_ha : map_has) {
+            std::string tmpStr = item::nname(map_ha.type) + _(" (nearby)");
             options.push_back(tmpStr);
         }
-        for( auto &player_ha : player_has ) {
-            options.push_back( item::nname( player_ha.type ) );
+        for (auto &player_ha : player_has) {
+            options.push_back(item::nname(player_ha.type));
         }
 
         if (options.empty()) { // This SHOULD only happen if cooking with a fire,
-            return;    // and the fire goes out.
+            selected.use_from = usage::use_from_none;
+            return selected;    // and the fire goes out.
         }
 
         // Get selection via a popup menu
-        size_t selection = menu_vec(false, _("Use which tool?"), options, hotkeys) - 1;
-        if (selection < map_has.size()) {
-            long quantity = map_has[selection].count * batch;
-            g->m.use_charges(pos3(), PICKUP_RANGE, map_has[selection].type, quantity );
+        int selection = menu_vec(can_cancel, _("Use which tool?"), options, hotkeys) - 1;
+        if (selection == -1) {
```

Comment:
Are you sure `menu_vec` returns 0 when canceled? The comment in output.h indicates otherwise:
```
// for the next two functions, if cancelable is true, esc returns the last option
```
If that comment is wrong, it should be corrected.

Also: I think there is a return or break messing after assigning `selected.use_from`. Currently it's always overridden with `use_from_map` or `use_from_player` a few lines below.